Add encoding and codepage params to win_command/win_shell (#54896)

h-hirokawa opened the pull request:
##### SUMMARY
Add new encoding and codepage params to win_command/win_shell. Now you can set custom encoding to stdout/stderr.
Since the default encoding is still UTF-8, so the PR has no breaking changes.

Fixes #54896 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
win_command
win_shell
Ansible.Process.cs
Ansible.ModuleUtils.CommandUtil.psm1

h-hirokawa commented:
@dagwieers @if-meaton @jborean93 @jhawkesworth @nitzmahone
The tests seems working properly, but shippable `T=windows/2008/1` env failed because the 45-minute time limit has been exceeded. Is there a workaround for this?

jborean93 commented:
@h-hirokawa, thanks for working on this. Currently we are in a freeze period for core modules and we probably won't be able to have a look at this PR for a few weeks while we are squashing bugs for the 2.8 release. I'll definitely be having a look at this when we start to look at core features again I just wanted to make sure that you are aware it may take some time for now.

ShachafGoldstein commented on lib/ansible/module_utils/csharp/Ansible.Process.cs:
```diff
@@ -300,6 +318,17 @@ public static Result CreateProcess(string command)
 
+            Encoding encInst = null;
+            int codepageInt = codepage.GetValueOrDefault(0);
+            if (codepageInt != 0)
```
Why do we need both encoding and codepage if eventually, you choose either?

ShachafGoldstein commented on lib/ansible/module_utils/csharp/Ansible.Process.cs:
```diff
@@ -300,6 +318,17 @@ public static Result CreateProcess(string command)
             FileStream stdinStream = new FileStream(stdinWrite, FileAccess.Write);
 
+            Encoding encInst = null;
+            int codepageInt = codepage.GetValueOrDefault(0);
```
This suppose to be `0`?

ShachafGoldstein commented:
Since you prefer codepage over encoding, maybe add a test that sets both and see that it returns the right one - i.e. set bad codepage and correct encoding and see that the output is bad

jborean93 commented:
@h-hirokawa really sorry for the delay on this one. If you haven't heard from me by the end of the week please ping me on this PR to remind me. I'm hoping to spin up a host with the Japanese culture settings to test out everything.

jborean93 commented:
So I was able to setup a test host to replicate this issue and I've played around with the changes here. I think we have a good set of changes here but it needs some slight tweaks. Ultimately we should only have 1 added parameter which sets the output encoding. This value can be any valid encoding `Name` based on the output of `[System.Text.Encoding]::GetEncodings()` and we can just default to `utf-8` to ensure backwards compatibility. I don't think we should also expose codepage as an alternative way of specifying the codepage because;

* The codepage is about the console output encoding that is used when functions use generic WriteLine functions. Having us set it to `utf-8` (`65001`) means we can mostly rely on utf-8 bytes being in the stream
* While we do set the codepage this is only valid for some cases, the common scenario relies on the underlying codepage already being set. This means we cannot rely on having a configurable codepage and shouldn't expose it
* This shouldn't be an issue unless the application that is run writes the characters bytes in their own encoding and ignores the console's codepage (this is what `ipconfig` seems to do)

So ultimately I think we should just change the PR to only have `encoding` parameter and I think we should have it set as `output_encoding` instead. This leaves us open to an `input_encoding` option in the future without being stuck in the existing UI.

jborean93 commented:
Just an update, because people shouldn't really have to touch this setting except for exceptional circumstances I propose we set the option to `output_encoding_override` and really hammer home in the docs that this shouldn't be touched unless it really needs to be done.  Having this set could lead to all sorts of weird behaviour as it typically shouldn't be required.

cc @nitzmahone

h-hirokawa commented:
@ShachafGoldstein @jborean93 
Thank you for your reviews and responses.
Unfortunately, I will be busy for the next few weeks, but then I will read your suggests carefully and work out for improvements.

h-hirokawa committed:
Add output_encoding_override params to win_command/win_shell (#54896)

This enhancement enables Ansible to parse the output of
localized commands that ignore the prompt code page.

h-hirokawa commented:
@jborean93 @ShachafGoldstein 
Sorry I kept you waiting.
After all, I have added only one `output_encoding_override` option to win_command/win_shell module.
Though some Shippable tests are still failing due to the 45 minutes limit, this seems to work correctly.

jborean93 committed:
Added changelog and minor nits

jborean93 commented:
@h-hirokawa thanks for your work on this one and sorry for the delay. I've made a commit which tidies a few things up and changes the test to output data similar to the `ipconfig` scenario that original brough up the issue. I've also added a changelog fragment so people can see what has changed in the 2.10 release when that comes out.

h-hirokawa commented:
@jborean93 Thanks a lot ! That’s very helpful, I can learn a lot from your additional implementation.
